File: Python/minDistance.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: minDistance.py
@Function: 编辑距离 动态规划
@Link: https://leetcode-cn.com/problems/edit-distance/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-16
"""


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        n = len(word1)
        m = len(word2)

        # 有一个字符串为空串
        if n * m == 0:
            return n + m

        # DP 数组
        D = [[0] * (m + 1) for _ in range(n + 1)]

        # 边界状态初始化
        for i in range(n + 1):
            D[i][0] = i
        for j in range(m + 1):
            D[0][j] = j

        # 计算所有 DP 值
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                left = D[i - 1][j] + 1
                down = D[i][j - 1] + 1
                left_down = D[i - 1][j - 1]
                if word1[i - 1] != word2[j - 1]:
                    left_down += 1
                D[i][j] = min(left, down, left_down)

        return D[n][m]


# --------------------------
if __name__ == "__main__":
    word1, word2 = "horse", "ros"
    # word1, word2 = "intention", "execution"

    solution = Solution()
    min_distance = solution.minDistance(word1, word2)
    print(f"The solution of this problem is {min_distance}")

File: C++/minDistance.cpp
/**
 * @File    : minDistance.cpp
 * @Brief   : 编辑距离 动态规划
 * @Link    : https://leetcode-cn.com/problems/edit-distance/
 * @Author  : Wei Li
 * @Date    : 2021-07-16
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class Solution
{
public:
    size_t minDistance(std::string word1, std::string word2)
    {
        size_t m = word1.length(), n = word2.length();
        std::vector<std::vector<size_t>> dp(m + 1, std::vector<size_t>(n + 1, 0));
        for (size_t i = 0; i <= m; ++i)
        {
            for (size_t j = 0; j <= n; ++j)
            {
                if (i == 0)
                {
                    dp[i][j] = j;
                }
                else if (j == 0)
                {
                    dp[i][j] = i;
                }
                else
                {
                    dp[i][j] = std::min(dp[i - 1][j - 1] + ((word1[i - 1] == word2[j - 1]) ? 0 : 1),
                                        std::min(dp[i - 1][j] + 1, dp[i][j - 1] + 1));
                }
            }
        }
        return dp[m][n];
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // std::string word1 = "horse";
    // std::string word2 = "ros";

    std::string word1 = "intention";
    std::string word2 = "execution";

    Solution solution;
    size_t min_distance = solution.minDistance(word1, word2);
    std::cout << "The solution of this problem is " << min_distance << std::endl;

    return 0;
}
